### ISSUE TITLE
Improve debugging experience for compiled code

### ISSUE THREAD

**Issue by robin-aws** [dafny-lang/dafny#1895]
We've observed Dafny users attempting to debug the C# code produced by the Dafny compiler, and encountering what looks like a runtime bug but is actually an artifact of the debugger evaluating functions in the wrong context. Credit to @cpitclaudel for all the lovely details below. :)

Example:

```dafny
include "../../../libraries/src/Wrappers.dfy"
import opened Wrappers

method test(n: int) returns (r: Result<string, string>) {
  if (n == 0) {
    return Failure("Zero!");
  }
  return Success("NonZero!");
}

method Main() {
  var t := test(0);
  match t {
    case Success(s) => print "OK: ", s;
    case Failure(s) => print "Failure: ", s;
  }
}
```

The resulting C# code, for the test method, looks like this:

```C#
    public static Wrappers_Compile._IResult<Dafny.ISequence<char>, Dafny.ISequence<char>> test(BigInteger n)
    {
      Wrappers_Compile._IResult<Dafny.ISequence<char>, Dafny.ISequence<char>> r = Wrappers_Compile.Result<Dafny.ISequence<char>, Dafny.ISequence<char>>.Default(Dafny.Sequence<char>.Empty);
      if ((n).Sign == 0) {
        r = Wrappers_Compile.Result<Dafny.ISequence<char>, Dafny.ISequence<char>>.create_Failure(Dafny.Sequence<char>.FromString("Zero!"));
        return r;
      }
      r = Wrappers_Compile.Result<Dafny.ISequence<char>, Dafny.ISequence<char>>.create_Success(Dafny.Sequence<char>.FromString("NonZero!"));
      return r;
      return r;
    }
```

Let's prettify it to make the discussion easier:

```C#
public static _IResult<DafnyString, DafnyString> test(BigInteger n) {
  var r = Result<DafnyString, DafnyString>.Default(DafnyString.Empty); // 1

  if (n.Sign == 0) { // 2
    r = Result<DafnyString, DafnyString>.create_Failure(Dafny.Sequence<char>.FromString("Zero!")); // 3
    return r;
  }

  r = Result<Dafny.ISequence<char>, Dafny.ISequence<char>>.create_Success(DafnyString.FromString("NonZero!"));
  return r;
}
```

In `// 1`, Dafny creates a local variable `r`; this is because the return parameter of the function is named `r`, and Dafny creates it as soon as we enter the function. In this case is not used in the rest of the function because all return paths reassign `r`.

Note that the `Default` for a `Result` is `Success`, so we create a `Success` value.

Next, the branch at `// 2` is taken, and in `// 3` we reassign `r`. Here `r` becomes a `Failure`.

Now if you open up a debugger and run to `//2`, you see `dtor_error = Exception of type System.InvalidCastException was thrown` in the variable watcher for the result variable `r`.

If you continue past `// 3`, you see `dtor_value = Exception of type System.InvalidCastException was thrown`.

These are both expected. The `dtor_*` properties are accessors that Dafny generates, but they are not safe to access unless the class is of the right instance. This is an artifact of the encoding that Dafny uses to simulate tagged unions; in fact I wrote a blog post about this style of implementation 10 years ago (!) at https://pit-claudel.fr/clement/blog/using-abstract-classes-to-simulate-tagged-unions-aka-sum-types/ .

Basically *all* instances of Result have both properties `dtor_value` and `dtor_error`, but it's only safe to access them if the instance is of the right type (`Result_Success` and `Result_Failure`, respectively).

Unfortunately the debugger does not know that it's unsafe to access these properties, so regardless of whether `r` is a `Success` or a `Failure` it evaluates both `dtor_value` (the one that only works for successes) and `dtor_error` (the one that only works for failures). This is why you're seeing these errors in the debugger, both for the `Success` `r` (upon entering the function) and the `Failure` `r` (within the branch before the return).

I'm sure there's an attribute somewhere that we could slap on these properties to tell the debugger that they are unsafe to evaluate.

**Comment by MikaelMayer** [comment]
[`[DebuggerBrowsable(DebuggerBrowsableState.Never)]`](https://lostechies.com/jamesgregory/2009/08/18/debugger-property-evaluation-side-effects/) ?

**Comment by cpitclaudel** [comment]
Yes, I have a patch for that

**Comment by cpitclaudel** [comment]
But it requires a refactoring of the AST writer in the compiler, too, or the generated code will look very bad ^^
